### ISSUE TITLE
socket.bind(Path(...)) does not work

### ISSUE THREAD

**Issue by socketpair** [python/cpython#99349]
```python
s = socket(AF_UNIX, SOCK_DGRAM)
s.bind(Path('/tmp/qwe3333'))
```

gives:

`TypeError: a bytes-like object is required, not 'PosixPath'`

According to some PEP all path-related functions should accept Path.

### Linked PRs
* gh-99599

**Comment by brettcannon** [comment]
> According to some PEP

I believe you're thinking of https://peps.python.org/pep-0519/ .



> all path-related functions should accept Path.

What https://peps.python.org/pep-0519/#standard-library-changes says is:

> It is expected that most APIs in Python’s standard library that currently accept a file system path will be updated appropriately to accept path objects (whether that requires code or simply an update to documentation will vary).

So "most", not "all". 😉 As such, I have updated this issue to be a feature request.

I don't know tons about sockets, but if you would like to submit a PR you probably need to find the code for `socket.bind()` and have it use `os.fsencode()` to get the bytes representation of the path.

**Comment by noamcohen97** [comment]
`socket.bind()` is currently accepting any bytes-like object and I wonder wether an instance that is both bytes-like and path-like should be converted using `bytes()` or `__fspath__()`?
Using `bytes()` will preserve the current behavior, but the same path-like object would be treated differently in `open()` and in `socket.bind()`. Therefore, I think `__fspath__()` should overcome `bytes() even though it would break the interface. WDTY?

**Comment by brettcannon** [comment]
> `socket.bind()` is currently accepting any bytes-like object and I wonder wether an instance that is both bytes-like and path-like should be converted using `bytes()` or `__fspath__()`?

That's actually [what `os.fsencode()` does](https://github.com/python/cpython/blob/8f18ac04d32515eab841172c956a8cb14bcee9c3/Lib/os.py#L806).

**Comment by noamcohen97** [comment]
`os.fsencode()` does not handle **any** bytes-like object like `socker.bind()` is currently accepting.

**Comment by socketpair** [comment]
1. It should accept Path. And use bytes(Path) to get bytes representation of address. It will use fsencode-like conversion.
2. It must also accept bare bytes() for abstract sockets.
3. It must accept bare str() for random paths and convert them using fsencode. Random symbols are not allowed anyway in this case. I mean zero bytes for example.

**Comment by noamcohen97** [comment]
> 3\. zero bytes for example

we do support null-bytes currently (https://github.com/python/cpython/blob/main/Modules/socketmodule.c#L1724)

**Comment by socketpair** [comment]
yes, eveything ok. Only one thing left - is to add Path() support.

**Comment by brettcannon** [comment]
> `os.fsencode()` does not handle **any** bytes-like object

Ah, sorry, misunderstood that you wanted the argument potentially passed through `bytes()` itself, not that you were specifically expecting the type.

**Comment by brettcannon** [comment]
As I said, I'm not a sockets expert. 😅 As such another core dev will need to take charge of this issue to see it through.
